src/modules/job-manager/purge.c

```diff
@@ -0,0 +1,381 @@
+{
+    flux_t *h = flux_future_get_flux (f_sync);
+    struct purge *purge = arg;
+
+    if (flux_future_get (f_sync, NULL) < 0) {
+        flux_log (h,
+                  LOG_ERR,
+                  "purge synchronization error: %s",
+                  future_strerror (f_sync, errno));
+    }
+    if (!(purge->f_purge)) {
+        flux_future_t *f;
+        if (!(f = purge_inactive_jobs (purge,
+                                       purge->age_limit,
+                                       purge->num_limit,
+                                       purge_batch_max))
+            || flux_future_then (f, -1, purge_continuation, purge) < 0) {
+            flux_future_destroy (f);
+            if (errno != ENODATA)
+                flux_log_error (h, "error creating purge KVS transaction");
+            goto done;
+        }
+        purge->f_purge = f;
+    }
+done:
+    flux_future_reset (f_sync);
+}
+
+/* Locate the message containing 'f'.
+ * Side effect: cursor is parked on this message, so flux_msglist_delete()
+ * can be used to delete it later.
+ */
+static const flux_msg_t *find_request (struct flux_msglist *l, flux_future_t *f)
+{
+    const flux_msg_t *msg;
+
+    msg = flux_msglist_first (l);
+    while (msg) {
+        if (flux_msg_aux_get (msg, "future") == f)
+            return msg;
+        msg = flux_msglist_next (l);
+    }
+    return NULL;
+}
+
+static void purge_request_continuation (flux_future_t *f, void *arg)
+{
+    flux_t *h = flux_future_get_flux (f);
+    struct purge *purge = arg;
+    int count = ptr2int (flux_future_aux_get (f, "count"));
+    const flux_msg_t *msg = find_request (purge->requests, f);
+
+    assert (msg != NULL);
+    if (flux_rpc_get (f, NULL) < 0) {
+        if (flux_respond_error (h, msg, errno, future_strerror (f, errno)) < 0)
+            flux_log_error (h, "error responding to purge request");
+        goto done;
+    }
+    if (flux_respond_pack (h, msg, "{s:i}", "count", count) < 0)
+        flux_log_error (h, "error responding to purge request");
+done:
+    // assumes cursor still positioned from find_request(), and destroys f
+    flux_msglist_delete (purge->requests);
+}
+
+static void purge_request_cb (flux_t *h,
+                              flux_msg_handler_t *mh,
+                              const flux_msg_t *msg,
+                              void *arg)
+{
+    struct purge *purge = arg;
+    double age_limit;
+    int num_limit;
+    int batch;
+    flux_future_t *f;
+    const char *errmsg = NULL;
+    flux_error_t error;
+
+    if (flux_request_unpack (msg,
+                             NULL,
+                             "{s:f s:i s:i}",
+                             "age_limit", &age_limit,
+                             "num_limit", &num_limit,
+                             "batch", &batch) < 0)
+        goto error;
+    if (age_limit != INACTIVE_AGE_UNLIMITED && age_limit < 0) {
+        errmsg = "age limit must be >= 0 (0=unlimited)";
```

Comment:
Is `errmsg` correct? (`INACTIVE_AGE_UNLIMITED` is `-1.` not `0`, but perhaps there's some subtlety I missed) FWIW, `0` would be a find value for unlimited